processing/src/main/java/io/druid/query/monomorphicprocessing/SpecializationService.java

```diff
@@ -0,0 +1,391 @@
+    }
+    catch (Exception e) {
+      throw new RuntimeException("Cannot access Unsafe methods", e);
+    }
+  }
+
+  /**
+   * If true, specialization is not actually done, an instance of prototypeClass is used as a "specialized" instance.
+   * Useful for analysis of generated assembly with JITWatch (https://github.com/AdoptOpenJDK/jitwatch), because
+   * JITWatch shows only classes present in the loaded JAR (prototypeClass should be), not classes generated during
+   * runtime.
+   */
+  private static final boolean fakeSpecialize = Boolean.getBoolean("fakeSpecialize");
+
+  /**
+   * Number of loop iterations, accounted via {@link SpecializationState#accountLoopIterations(long)} in
+   * {@link WindowedLoopIterationCounter} during the last hour window, after which WindowedLoopIterationCounter decides
+   * to specialize class for the specific runtimeShape. The default value is chosen to be so that the specialized
+   * class will likely be compiled with C2 HotSpot compiler with the default values of *BackEdgeThreshold options.
+   */
+  private static final long triggerSpecializationIterationsThreshold =
+      Integer.getInteger("triggerSpecializationIterationsThreshold", 10_000);
+
+  private static final ExecutorService classSpecializationExecutor = Execs.singleThreaded("class-specialization-%d");
+
+  private static final AtomicLong specializedClassCounter = new AtomicLong();
+
+  private static final ClassValue<PerPrototypeClassState> perPrototypeClassState =
+      new ClassValue<PerPrototypeClassState>()
+      {
+        @Override
+        protected PerPrototypeClassState computeValue(Class<?> type)
+        {
+          return PerPrototypeClassState.of(type);
+        }
+      };
+
+  /**
+   * @param <T> type of query processing algorithm
+   * @see SpecializationService class-level javadoc for details
+   */
+  public static <T> SpecializationState<T> getSpecializationState(
+      Class<? extends T> prototypeClass,
+      String runtimeShape
+  )
+  {
+    return getSpecializationState(prototypeClass, runtimeShape, ImmutableMap.<Class<?>, Class<?>>of());
+  }
+
+  /**
+   * @param classRemapping classes, that should be replaced in the bytecode of the given prototypeClass when specialized
+   * @see #getSpecializationState(Class, String)
+   */
+  @SuppressWarnings("unchecked")
+  public static <T> SpecializationState<T> getSpecializationState(
+      Class<? extends T> prototypeClass,
+      String runtimeShape,
+      ImmutableMap<Class<?>, Class<?>> classRemapping
+  )
+  {
+    return perPrototypeClassState.get(prototypeClass).getSpecializationState(runtimeShape, classRemapping);
+  }
+
+  static class PerPrototypeClassState<T>
+  {
+    static <T> PerPrototypeClassState<T> of(Class<T> prototypeClass)
+    {
+      return new PerPrototypeClassState<>(prototypeClass);
+    }
+
+    private final Class<T> prototypeClass;
+    private final ConcurrentMap<SpecializationId, SpecializationState<T>> specializationStates =
```

Comment:
Theoretically this map could grow as large as many query runtime shapes (combinations) with 1-2 aggregators are possible in Druid, it's hard to calculate, probably about 1000. In practice I decided not to overengineer until this hits anybody. In our production this map has ~ 20 entries.